File: recursion/c002.cpp
#include<iostream>
#include<vector>
#include<string>
using namespace std;

int permutations_void(string s,string ans){
    if(s.size()==0){
        cout<<ans<<endl;
        return 1;
    }
    int count=0;
    for(int  i=0;i<s.size();i++){
        string smallstr= s.substr(0,i) + s.substr(i+1);
        count += permutations_void(smallstr,ans+s[i]);
    }
    return count;
}

vector<string> permutations_ret(string s){    
    vector<string> ans;
    if(s.size()==1){
        ans.push_back(s);
        return ans;
    }
    char ch=s[0];
    vector<string> smallans=permutations_ret(s.substr(1));
    for(string str:smallans){
        for(int i=0;i<=str.size();i++){
            ans.push_back(str.substr(0,i)+ch+str.substr(i));
        }
    }
    return ans;
}
int main(){
    string s="abc";
    //cout<<permutations_void(s,"")<<endl;
    vector<string> ans=permutations_ret("abc");
    for(int i=0;i<s.size();i++){
        cout<<ans[i]<<" ";
    }
    return 0;
}

File: recursion/c002.java
//  Permutations of a String (all cases)

import java.util.*;
public class c002 {
    public static void main(String[]args){
        String s="abcd";
        Permutations_without_dupli01(s,"");
        //System.out.println(ans);
    }

    public static ArrayList<String> permutations_with_dupli_ret(String s) {
        if(s.length()==1){
            ArrayList<String> ans=new ArrayList<String>();
            ans.add(s);
            return ans;
        }
        char ch=s.charAt(0);
        ArrayList<String> smallans=new ArrayList<String>();
        ArrayList<String> ans=new ArrayList<String>();
        smallans=permutations_with_dupli_ret(s.substring(1));
        for(String ns:smallans){
            for(int i=0;i<=ns.length();i++){
                String a=ns.substring(0,i)+ch+ns.substring(i);
                ans.add(a);
            }            
        }
        return ans;
    } 
    
    public static void permutations_with_dupli_void(String str,String ans){
        if(str.length()==0){
            System.out.println(ans);
            return;
        }

        for(int i=0;i<str.length();i++){
            char ch=str.charAt(i);
            String smallstr=str.substring(0,i)+str.substring(i+1);
            permutations_with_dupli_void(smallstr, ans+ch);
        }
        return;
    }


    public static void Permutations_without_dupli01(String s,String ans){
        if(s.length()==0){
            System.out.println(ans);
            return;
        }
        int visited[] =new int[26];
        for(int i=0;i<s.length();i++){
            visited[i]=0;
        }
        for(int i=0;i<s.length();i++){
            char ch=s.charAt(i);
            int idx=ch-'a';            
            if(visited[idx]==0){
                visited[idx]++;
                String smallstr=s.substring(0,i)+s.substring(i+1);
                Permutations_without_dupli01(smallstr, ans+ch);
            }
        }
        return;
    } 
}